src/picongpu/include/particles/synchrotronPhotons/PhotonCreator.hpp

```diff
@@ -0,0 +1,319 @@
+#include "traits/frame/GetCharge.hpp"
+
+#include "random/methods/XorMin.hpp"
+#include "random/distributions/Uniform.hpp"
+#include "random/RNGProvider.hpp"
+
+#include "traits/Resolve.hpp"
+#include "mappings/kernel/AreaMapping.hpp"
+
+#include "fields/FieldB.hpp"
+#include "fields/FieldE.hpp"
+
+#include "compileTime/conversion/TypeToPointerPair.hpp"
+#include "memory/boxes/DataBox.hpp"
+
+namespace picongpu
+{
+namespace particles
+{
+namespace synchrotronPhotons
+{
+
+/** Functor creating photons from electrons according to synchrotron radiation.
+ *
+ * The numerical model is taken from:
+ *
+ * Gonoskov, A., et al. "Extended particle-in-cell schemes for physics
+ * in ultrastrong laser fields: Review and developments."
+ * Physical Review E 92.2 (2015): 023305.
+ *
+ * This functor is called by the general particle creation module.
+ *
+ * \tparam T_ElectronSpecies
+ * \tparam T_PhotonSpecies
+ */
+template<typename T_ElectronSpecies, typename T_PhotonSpecies>
+struct PhotonCreator
+{
+    typedef T_ElectronSpecies ElectronSpecies;
+    typedef T_PhotonSpecies PhotonSpecies;
+
+    typedef typename ElectronSpecies::FrameType FrameType;
+
+    /* specify field to particle interpolation scheme */
+    typedef typename PMacc::traits::Resolve<
+        typename GetFlagType<FrameType,interpolation<> >::type
+        >::type Field2ParticleInterpolation;
+
+    /* margins around the supercell for the interpolation of the field on the cells */
+    typedef typename GetMargin<Field2ParticleInterpolation>::LowerMargin LowerMargin;
+    typedef typename GetMargin<Field2ParticleInterpolation>::UpperMargin UpperMargin;
+
+    /* relevant area of a block */
+    typedef SuperCellDescription<
+        typename MappingDesc::SuperCellSize,
+        LowerMargin,
+        UpperMargin
+        > BlockArea;
+
+    BlockArea BlockDescription;
+
+    typedef MappingDesc::SuperCellSize TVec;
+
+    typedef FieldE::ValueType ValueType_E;
+    typedef FieldB::ValueType ValueType_B;
+
+private:
+    /* global memory EM-field device databoxes */
+    PMACC_ALIGN(eBox, FieldE::DataBoxType);
+    PMACC_ALIGN(bBox, FieldB::DataBoxType);
+    /* shared memory EM-field device databoxes */
+    PMACC_ALIGN(cachedE, DataBox<SharedBox<ValueType_E, typename BlockArea::FullSuperCellSize,1> >);
+    PMACC_ALIGN(cachedB, DataBox<SharedBox<ValueType_B, typename BlockArea::FullSuperCellSize,0> >);
+
+    PMACC_ALIGN(curF_1, SynchrotronFunctions::SyncFuncCursor);
+    PMACC_ALIGN(curF_2, SynchrotronFunctions::SyncFuncCursor);
+
+    PMACC_ALIGN(photon_mom, float3_X);
+
+    /* random number generator */
+    typedef PMacc::random::RNGProvider<simDim, PMacc::random::methods::XorMin> RNGFactory;
+    typedef PMacc::random::distributions::Uniform<float> Distribution;
+    typedef typename RNGFactory::GetRandomType<Distribution>::type RandomGen;
+    RandomGen randomGen;
+
+public:
+    /* host constructor initializing member : random number generator */
+    PhotonCreator(
```

Comment:
The name of this methods does not suggest that it is just a initialization. Please think of a more suitable name.